src/we_aes_gcm.c

```diff
@@ -356,153 +302,142 @@ static int we_aes_gcm_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
-            if (rc != 0) {
-                WOLFENGINE_ERROR_FUNC(WE_LOG_CIPHER, "wc_AesGcmEncrypt_ex", rc);
-                ret = -1;
-            }
-            if (ret == 1) {
-                WOLFENGINE_MSG_VERBOSE(WE_LOG_CIPHER, "Encrypted %zu bytes "
-                                       "(AES-GCM):", len);
-                WOLFENGINE_BUFFER(WE_LOG_CIPHER, out, (unsigned int)len);
-                WOLFENGINE_MSG_VERBOSE(WE_LOG_CIPHER, "AES-GCM tag:");
-                WOLFENGINE_BUFFER(WE_LOG_CIPHER, aes->tag, aes->tagLen);
-            }
-        }
-        else if (ret == 1) {
-            WOLFENGINE_MSG(WE_LOG_CIPHER, "AES-GCM Decrypting");
-        #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-            if (aes->tagLen == 0) {
-                WOLFENGINE_MSG_VERBOSE(WE_LOG_CIPHER, "No tag yet, deferring "
-                    "decryption until \"final\" called with tag.");
-                if (aes->decryptBuf != NULL) {
-                    OPENSSL_free(aes->decryptBuf);
-                }
-                aes->decryptBuf = (unsigned char*)OPENSSL_malloc(len);
-                if (aes->decryptBuf == NULL) {
-                    WOLFENGINE_ERROR_FUNC_NULL(WE_LOG_CIPHER,
-                        "OPENSSL_malloc", aes->decryptBuf);
+            if (out == NULL && len != 0 && in != NULL) {
+                WOLFENGINE_MSG(WE_LOG_CIPHER, "Resizing stored AAD and "
+                    "appending data, len = %d", (int)len);
+                /* Resize stored AAD and append new data. */
+                p = (unsigned char*)OPENSSL_realloc(aes->aad, aes->aadLen +
+                        (int)len);
+                if (p == NULL) {
+                    WOLFENGINE_ERROR_FUNC_NULL(WE_LOG_CIPHER, "OPENSSL_realloc",
+                        p);
                     ret = -1;
                 }
                 else {
-                    XMEMCPY(aes->decryptBuf, in, len);
-                    aes->decryptBufLen = len;
+                    /* Copy in new data after existing data. */
+                    aes->aad = p;
+                    XMEMCPY(aes->aad + aes->aadLen, in, len);
+                    aes->aadLen += len;
+                    ret = (int)len;
                 }
-                /* No data decrypted, yet, so return 0. */
-                ret = 0;
-            }
-            else
-        #endif
-            {
-                ret = we_aes_gcm_decrypt(aes, out, in, len);
             }
-        }
-
-    #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-        if (aes->enc || (!aes->enc && aes->tagLen > 0))
-    #endif
-        {
-            /* Dispose of any AAD - all used now. */
-            OPENSSL_free(aes->aad);
-            aes->aad = NULL;
-            aes->aadLen = 0;
-        }
-        if (ret == 1) {
-            if (aes->ivInc) {
-                int i;
-                for (i = aes->ivLen - 1; i >= aes->ivLen - 8; i--) {
-                    if ((++aes->iv[i]) != 0) {
-                        break;
+            else {
```

Comment:
We only want to go here if out != NULL right?

Reply:
out can be NULL in the GMAC case:
```
Breakpoint 2, we_aes_gcm_cipher (ctx=0x5555555976e0, out=0x0, in=0x0, len=0) at src/we_aes_gcm.c:330
330	                if (!aes->ivSet) {
(gdb) where
#0  we_aes_gcm_cipher (ctx=0x5555555976e0, out=0x0, in=0x0, len=0) at src/we_aes_gcm.c:330
#1  0x00007ffff7d5d4b3 in EVP_DecryptFinal_ex (ctx=0x5555555976e0, out=0x0, outl=0x7fffffffd024) at crypto/evp/evp_enc.c:539
#2  0x000055555555ab59 in test_aes_tag_dec (e=0x5555555900e0, cipher=0x7ffff7eb8980 <aes_128_gcm>, key=0x7fffffffd0e0 "", iv=0x7fffffffd0c0 "", ivLen=12, aad=0x7fffffffd0bc "AAD", msg=0x0, 
    len=0, enc=0x0, tag=0x7fffffffd0d0 "'V\026\064\233\261\347vk\026:Ì\262Sx", dec=0x0, ccm=0, ccmL=0) at test/test_aestag.c:181
#3  0x000055555555aff5 in test_aes_gcm_gmac (e=0x5555555900e0, data=0x0, cipher=0x7ffff7eb8980 <aes_128_gcm>, keyLen=16, ivLen=12) at test/test_aestag.c:270
#4  0x000055555555be76 in test_aes128_gcm (e=0x5555555900e0, data=0x0) at test/test_aestag.c:582
#5  0x0000555555569c94 in run_tests (e=0x5555555900e0, runAll=1) at test/unit.c:541
#6  0x000055555556a3c9 in main (argc=0, argv=0x7fffffffd2e0) at test/unit.c:735
```

Reply:
Right! Thanks.